### PR TITLE
Extend optical flow tutorial

### PR DESCRIPTION
**Pull request by mehlukas**
Use the [python optical flow tutorial](https://github.com/opencv/opencv/blob/master/doc/py_tutorials/py_video/py_lucas_kanade/py_lucas_kanade.markdown) as a base to create a general optical flow tutorial. This is done according to the [Document improvement plan](https://github.com/opencv/opencv/wiki/Documentation-improvement-plan) and the discussion in #9520.
I tried to create a similar result as in the [background subtraction tutorial](https://github.com/opencv/opencv/blob/master/doc/tutorials/video/background_subtraction/background_subtraction.markdown).

<cut/>

### This pullrequest changes

* extract python example code from [python tutorial on optical flow](https://github.com/opencv/opencv/blob/master/doc/py_tutorials/py_video/py_lucas_kanade/py_lucas_kanade.markdown) into the proper subfolder in samples
* add C++ example code
* add tutorial to the common tutorial folder with python and C++ code examples that can be toggled.

[Documentation link](http://pullrequest.opencv.org/buildbot/export/pr/14314/docs/d4/dee/tutorial_optical_flow.html).

```
force_builders=linux,windows,mac,docs
test_modules=none
build_contrib:Win64=OFF
build_contrib:Win64 OpenCL=OFF
```

### CODE REVIEW

File: samples/cpp/tutorial_code/video/optical_flow/optical_flow_dense.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        Mat flow(prvs.size(), CV_32FC2);
+        calcOpticalFlowFarneback(prvs, next, flow, 0.5, 3, 15, 3, 5, 1.2, 0);
+
+        // visualization
+        Mat flow_parts[2];
+        split(flow, flow_parts);
+        Mat magnitude, angle, magn_norm;
+        cartToPolar(flow_parts[0], flow_parts[1], magnitude, angle, true);
+        normalize(magnitude, magn_norm, 0.0f, 1.0f, NORM_MINMAX);
+        angle = angle / 360.0 * 255.0;
```
Comment (llschloesser) [inline samples/cpp/tutorial_code/video/optical_flow/optical_flow_dense.cpp]: @mehlukas 
Is this correct? I see an artificial red/blue boundary that doesn't match the flow vectors. Remember that OpenCV expects HSV hue as 0-180. The following has no red/blue boundary issue:

```
void flowViz(const cv::Mat& flow, cv::Mat& dst)
{
  cv::Mat uv[2];
  cv::split(flow, uv);
  cv::Mat magnitude, angle;
  cv::cartToPolar(uv[0], uv[1], magnitude, angle, true); 
  cv::normalize(magnitude, magnitude, 0.0, 1.0, cv::NORM_MINMAX);
  angle *= ((1.f / 360.f) * (180.f / 255.f));

  cv::Mat hsv_[3], hsv, hsv8;
  hsv_[0] = angle;
  hsv_[1] = cv::Mat::ones(angle.size(), CV_32F);
  hsv_[2] = magnitude;
  cv::merge(hsv_, 3, hsv);
  hsv.convertTo(hsv8, CV_8U, 255.0); 
  cv::cvtColor(hsv8, dst, cv::COLOR_HSV2BGR);
}
```

Comment (mehlukas) [reply samples/cpp/tutorial_code/video/optical_flow/optical_flow_dense.cpp]: Yes, you're right! Thanks for noticing.

File: samples/cpp/tutorial_code/video/optical_flow/optical_flow_dense.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#include <opencv2/imgproc.hpp>
+#include <opencv2/videoio.hpp>
+#include <opencv2/video.hpp>
+
+using namespace cv;
+using namespace std;
+
+int main()
+{
+    VideoCapture capture("vtest.avi");
```
Comment (alalek) [inline samples/cpp/tutorial_code/video/optical_flow/optical_flow_dense.cpp]: Please use this:
- `samples::findFile("vtest.avi")`

File: samples/cpp/tutorial_code/video/optical_flow/optical_flow.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+#include <opencv2/imgproc.hpp>
+#include <opencv2/videoio.hpp>
+#include <opencv2/video.hpp>
+
+using namespace cv;
+using namespace std;
+
+int main()
+{
+    VideoCapture capture("slow_traffic_small.mp4");
```
Comment (alalek) [inline samples/cpp/tutorial_code/video/optical_flow/optical_flow.cpp]: Where is located this file?

Comment (mehlukas) [reply samples/cpp/tutorial_code/video/optical_flow/optical_flow.cpp]: This file which is also named "slow.flv" in the [python tutorial](https://github.com/opencv/opencv/blob/master/doc/py_tutorials/py_video/py_lucas_kanade/py_lucas_kanade.markdown) has actually been requested/missing multiple times in the forum, e.g. [here](http://answers.opencv.org/question/43986/python-tutorial-sample-source-file/), [here](http://answers.opencv.org/question/70166/download-example-file/) or [here](http://answers.opencv.org/question/102258/where-to-get-slowflv/). How do we deal with this?
PS: Is there a reason the opencv buildbot seems to be stuck at "Waiting for builds"?

Comment (alalek) [reply samples/cpp/tutorial_code/video/optical_flow/optical_flow.cpp]: Agree.

Can we pass video via command line parameter?
We can show URL link for proposed video via "usage help". [Example](https://github.com/opencv/opencv/blob/4.1.0/samples/dnn/colorization.cpp#L51-L53).

**Pull request merged**